---
title: "tenXplore: ontology for scRNA-seq, applied to 10x 1.3 million neurons"
author: "Vincent J. Carey, stvjc at channing.harvard.edu"
date: "`r format(Sys.time(), '%B %d, %Y')`"
vignette: >
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteIndexEntry{tenXplore: ontology for scRNA-seq, applied to 10x 1.3 million neurons}
  %\VignetteEncoding{UTF-8}
output:
  BiocStyle::pdf_document:
    toc: yes
    number_sections: yes
  BiocStyle::html_document:
    highlight: pygments
    number_sections: yes
    theme: united
    toc: yes
---

```{r setupp,echo=FALSE,results="hide"}
suppressWarnings({
suppressPackageStartupMessages({
library(tenXplore)
library(org.Mm.eg.db)
library(org.Hs.eg.db)
})
})
```

# Introduction

This package is intended to illustrate the role of formal
ontology in the analysis of single-cell RNA-seq experiments.

We anticipate that both unsupervised and supervised methods
of statistical learning
will be useful.  

- The [10x 1.3 million neuron dataset](https://community.10xgenomics.com/t5/10x-Blog/Our-1-3-million-single-cell-dataset-is-ready-to-download/ba-p/276) primarily
invites unsupervised analysis, as no anatomical or other
structural details are provided along with single-cell expression
measures.  
- Other studies that include details on anatomic source
or other non-transcriptomic features of cellular identity
would be amenable to supervised analysis, which will
benefit from standardization of metadata about experimental
factors and anatomic structures.

## An enumeration of cell types

<!--
label_URL("<http://www.ebi.ac.uk/efo/EFO_0000324>", getModel(ee), getWorld(ee))
$c
[1] "\"cell type\""
-->

We used the [Experimental Factor Ontology](https://www.ebi.ac.uk/efo/) 'cell type' class ([EFO_0000324](https://www.ebi.ac.uk/ols/ontologies/efo/terms?iri=http%3A%2F%2Fwww.ebi.ac.uk%2Fefo%2FEFO_0000324)) to obtain
an enumeration of cell types.  As of August 22 2017 it is an open question
whether [Cell Ontology](http://obofoundry.org/ontology/cl.html) should be used for this purpose.  The author's
subjective impression is that EFO has a simpler collection of terms for cell types,
while Cell Ontology has a better collection of terms for types of neurons.

## Basic operations based on redland with OWL

This package ships with the OWL representation of the [Cell Ontology](http://obofoundry.org/ontology/cl.html).
This is imported as RDF using `r CRANpkg("redland")`.  
```{r useCO}
library(tenXplore)
co = buildCellOntSupport()
co
```
At this time,
elementary manipulations of
the ontology involve collecting the children, siblings, or labels for given
URIs.
```{r useCO2}
cochil = children_URL("<http://purl.obolibrary.org/obo/CL_0000540>", 
    model=getModel(co), world=getWorld(co))
cochil
label_URL("<http://purl.obolibrary.org/obo/CL_0000540>", 
    model=getModel(co), world=getWorld(co))
siblings_URL("<http://purl.obolibrary.org/obo/CL_0000540>", 
    model=getModel(co), world=getWorld(co))
```

# Application: finding genes annotated to neuron subtypes

We focus on mouse.  The neuron subtypes identified as
OWL subclasses of "neuron" have names
```{r getcl}
cleanNames = function(tset) {
 tset@cleanFrame$clean
}
cleanNames(cochil)
```

We would like to see if the expression data would allow us to discriminate
neurons of these different types.

## Bridging from Cell Ontology to mouse genes

There is no formal linkage at present between terms
of Cell Ontology and those of Gene Ontology.  Research
on inference of tissue of origin from expression 
signatures has led to accurate classifiers (Lee, Krishnan, Troyanskaya) and
applications in cell mixture deconvolution (Houseman).  
Formal work in ontology bridging has been described but the
specific task of mapping from Cell Ontology terms
to Gene Ontology terms has not culminated in any
programmatically available resource.

We apply approximate pattern matching (agrep in R) to
find gene ontology terms that are apparently relevant to
cell type vocabulary terms of interest.  These are then
mapped to gene annotation.  Simple (non-vectorized)
functions that accomplish this in an organism-specific
are straightforward using the OrgDb packages.  We
serialized all GO terms for convenience with this package,
in the data object `allGOterms`.

```{r lkfuns}
cellTypeToGO = function(celltypeString, ...) {
 allGOterms[agrep(celltypeString, allGOterms[,2],...),]
 }
cellTypeToGenes = function(celltypeString, orgDb, cols=c("ENSEMBL", "SYMBOL"),...) {
 g = cellTypeToGO(celltypeString, ...)
 na.omit(select(orgDb, keys=g$GOID, keytype="GO", columns=cols))
}
library(tenXplore)
data(allGOterms)
cellTypeToGO("serotonergic neuron")
cellTypeToGenes("serotonergic neuron", orgDb=org.Mm.eg.db)
cellTypeToGenes("serotonergic neuron", orgDb=org.Hs.eg.db)
```

## Discrimination of neuron types: exploratory multivariate analysis



